File: python_service/voice_processor.py
import math

# ==========================================
# 1. SHARED CORE: Acoustic Quantizer
# ==========================================
class AcousticQuantizer:
    """
    音響数値をLLMが解釈可能な「文脈タグ」に変換するクラス。
    0-100の数値を5段階の形容詞にマッピングし、レポートの表現力を底上げする。
    """

    @staticmethod
    def get_pitch_tag(value):
        if value < 20: return "Sub-bass / Gravitas (重厚)"
        if value < 40: return "Deep / Resonant (共鳴)"
        if value < 60: return "Baritone / Grounded (安定的)"
        if value < 80: return "Tenor / Clear (明瞭)"
        return "Soprano / Crystalline (透き通った)"

    @staticmethod
    def get_speed_tag(value):
        if value < 20: return "Largo / Contemplative (熟考)"
        if value < 40: return "Andante / Deliberate (慎重)"
        if value < 60: return "Moderato / Conversational (会話的)"
        if value < 80: return "Allegro / Energetic (活発)"
        return "Presto / Urgent (性急)"

    @staticmethod
    def get_volume_tag(value):
        if value < 20: return "Whisper / Intimate (親密・秘密)"
        if value < 40: return "Soft / Gentle (穏やか・配慮)"
        if value < 60: return "Mezzo / Balanced (バランス)"
        if value < 80: return "Forte / Projecting (発信・主張)"
        return "Fortissimo / Commanding (威厳・支配)"

    @staticmethod
    def get_tone_tag(value):
        # Tone: 低い＝ハスキー/雑味、高い＝クリア/純音
        if value < 20: return "Husky / Textured (陰影・ハスキー)"
        if value < 40: return "Smoky / Warm (温かみ・スモーキー)"
        if value < 60: return "Neutral / Natural (自然的)"
        if value < 80: return "Bright / Polished (磨かれた)"
        return "Piercing / Pure (純粋・鋭い)"

    @classmethod
    def get_all_tags(cls, p, s, v, t):
        return {
            "Pitch_Tag": cls.get_pitch_tag(p),
            "Speed_Tag": cls.get_speed_tag(s),
            "Volume_Tag": cls.get_volume_tag(v),
            "Tone_Tag": cls.get_tone_tag(t)
        }

# ==========================================
# 2. SOLO ENGINE: Voice Identity & Gap Analysis
# ==========================================
class SoloIdentityEngine:
    """
    単身者向け($10プラン)。
    「声のアーキタイプ」を特定し、ユーザー申告のMBTIとの「ギャップ」を計算する。
    """
    
    def __init__(self, p, s, v, t, mbti_type):
        self.p = p
        self.s = s
        self.v = v
        self.t = t
        self.mbti = mbti_type.upper()

    def _calculate_axes(self):
        # Axis 1: Projection (投射力) - 外交的エネルギー
        # VolumeとSpeedが高いほど、前に出る力が強い
        projection = (self.v + self.s) / 2
        
        # Axis 2: Texture (質感) - 情報処理の温度感
        # Toneが高い(Clear)かつPitchが高い = Cool/Intellectual
        # Toneが低い(Husky)かつPitchが低い = Warm/Emotional
        texture = (self.p + self.t) / 2
        
        return projection, texture

    def _determine_archetype(self, proj, text):
        # 4象限マトリクスによる分類
        if proj >= 50:
            if text >= 50:
                return {
                    "Label": "The Lightning Rod (雷撃の扇動者)",
                    "Quote": "Words like electric shocks. Impossible to ignore.",
                    "Vibe": "High Voltage & Sharp"
                }
            else:
                return {
                    "Label": "The Thunder King (轟く覇者)",
                    "Quote": "A voice that shakes the floorboards. Pure dominance.",
                    "Vibe": "High Voltage & Heavy"
                }
        else:
            if text >= 50:
                return {
                    "Label": "The Ice Sculptor (氷の彫刻家)",
                    "Quote": "Precision over volume. Every syllable cuts deep.",
                    "Vibe": "Low Voltage & Sharp"
                }
            else:
                return {
                    "Label": "The Midnight FM (深夜のDJ)",
                    "Quote": "Velvet frequencies. You bypass ears and speak to the soul.",
                    "Vibe": "Low Voltage & Heavy"
                }

    def _analyze_gap(self, proj, text):
        # MBTIから期待される音声値を推論
        # E vs I -> Expected Projection
        expected_proj = 80 if 'E' in self.mbti else 30
        # T vs F -> Expected Texture (T=Cool/High, F=Warm/Low)
        expected_text = 80 if 'T' in self.mbti else 30

        proj_gap = proj - expected_proj
        text_gap = text - expected_text
        
        tags = []
        
        # Gap Diagnosis logic
        if proj_gap > 30: tags.append("Over-Amplified (Masking Introversion)")
        elif proj_gap < -30: tags.append("Under-Projecting (Holding Back)")
        else: tags.append("Authentic Projection (Aligned)")

        if text_gap > 30: tags.append("Cooler than Personality (Logical Mask)")
        elif text_gap < -30: tags.append("Warmer than Personality (Social Mask)")
        else: tags.append("Authentic Texture (Aligned)")

        return {
            "Projection_Delta": int(proj_gap),
            "Texture_Delta": int(text_gap),
            "Diagnosis_Tags": tags
        }

    def generate_payload(self):
        proj, text = self._calculate_axes()
        archetype = self._determine_archetype(proj, text)
        gap_analysis = self._analyze_gap(proj, text)
        quantized_metrics = AcousticQuantizer.get_all_tags(self.p, self.s, self.v, self.t)

        return {
            "User_MBTI": self.mbti,
            "Voice_Archetype": {
                "Label": archetype["Label"],
                "Quote": archetype["Quote"],
                "Stats": {"Projection": int(proj), "Texture": int(text)}
            },
            "Gap_Analysis": gap_analysis,
            "Raw_Metrics_Tags": quantized_metrics,
            "System_Instruction": "Focus on the discrepancy between MBTI and Voice Archetype."
        }

# ==========================================
# 3. COUPLE ENGINE: Resonance & SCM Analysis
# ==========================================
class CoupleResonanceEngine:
    """
    カップル向け($15プラン)。
    SCM(Stereotype Content Model)と音響シンクロ率を計算し、関係性をプロファイリングする。
    """
    
    # 職業別ベーススコア (Competence, Warmth)
    JOB_DB = {
        "lawyer": (90, 30), "executive": (90, 20), "engineer": (85, 40),
        "doctor": (85, 70), "founder": (85, 65), "consultant": (80, 40),
        "artist": (50, 90), "teacher": (60, 85), "designer": (60, 80),
        "nurse": (50, 95), "writer": (40, 80), "musician": (45, 90),
        "student": (50, 60), "sales": (70, 80), "other": (50, 50)
    }

    def __init__(self, user_a_data, user_b_data):
        # user_data expects: {'name': str, 'job': str, 'accent': str, 'p': int, 's': int, 'v': int, 't': int}
        self.ua = user_a_data
        self.ub = user_b_data

    def _calculate_scm(self, job, p, s, v, t):
        # 1. 職業ベース値
        base_c, base_w = self.JOB_DB.get(job.lower(), (50, 50))
        
        # 2. 音声による補正 (Dynamic Adjustment)
        # Competence: 速く(S)、大きい(V)声は有能に見える
        voice_c_impact = ((s + v) / 2 - 50) * 0.5
        
        # Warmth: トーンが高く(T:Clear)、ピッチが高すぎない声は温かい
        # ※ピッチが高すぎる(>85)とキンキンしてWarmthが下がる調整を入れる
        p_factor = p if p < 85 else (170 - p) 
        voice_w_impact = ((t + p_factor) / 2 - 50) * 0.5

        final_c = max(0, min(100, base_c + voice_c_impact))
        final_w = max(0, min(100, base_w + voice_w_impact))
        
        # 3. ペルソナラベル付与
        if final_c >= 50 and final_w >= 50: label = "The Charismatic Ideal (Admiration)"
        elif final_c >= 50 and final_w < 50: label = "The Efficient Strategist (Respect)"
        elif final_c < 50 and final_w >= 50: label = "The Empathetic Soul (Sympathy)"
        else: label = "The Free Spirit (Unconventional)"
        
        return {"Competence": int(final_c), "Warmth": int(final_w), "Archetype": label}

    def _calculate_synergy(self):
        # 全指標の差分(Delta)を計算
        delta_p = abs(self.ua['p'] - self.ub['p'])
        delta_s = abs(self.ua['s'] - self.ub['s'])
        delta_v = abs(self.ua['v'] - self.ub['v'])
        delta_t = abs(self.ua['t'] - self.ub['t'])
        
        mean_delta = (delta_p + delta_s + delta_v + delta_t) / 4
        sync_score = 100 - mean_delta # シンクロ率 (高いほど似ている)

        # Dynamic Typeの判定
        if sync_score > 80:
            dtype = "High-Sync / Mirroring"
            desc = "Like looking in an acoustic mirror."
        elif sync_score < 40:
            dtype = "High-Contrast / Opposites"
            desc = "Magnetic attraction of opposites."
        else:
            dtype = "Complementary / Balanced"
            desc = "A healthy mix of similarity and difference."
            
        return {
            "Sync_Score": int(sync_score),
            "Mean_Delta": int(mean_delta),
            "Dynamic_Type": dtype,
            "Description": desc,
            "Volume_Delta_Specific": self.ua['v'] - self.ub['v'] # 正ならAが大きい
        }

    def generate_payload(self):
        # Process User A
        scm_a = self._calculate_scm(self.ua['job'], self.ua['p'], self.ua['s'], self.ua['v'], self.ua['t'])
        tags_a = AcousticQuantizer.get_all_tags(self.ua['p'], self.ua['s'], self.ua['v'], self.ua['t'])
        
        # Process User B
        scm_b = self._calculate_scm(self.ub['job'], self.ub['p'], self.ub['s'], self.ub['v'], self.ub['t'])
        tags_b = AcousticQuantizer.get_all_tags(self.ub['p'], self.ub['s'], self.ub['v'], self.ub['t'])

        # Synergy
        synergy = self._calculate_synergy()

        return {
            "Report_Type": "Couple_Resonance_v1",
            "Relationship_Core": synergy,
            "User_A_Insight": {
                "Name": self.ua['name'],
                "Profile": f"{self.ua['job']} ({self.ua['accent']})",
                "SCM_Profile": scm_a,
                "Acoustic_Tags": tags_a
            },
            "User_B_Insight": {
                "Name": self.ub['name'],
                "Profile": f"{self.ub['job']} ({self.ub['accent']})",
                "SCM_Profile": scm_b,
                "Acoustic_Tags": tags_b
            },
            "Narrative_Hint": f"User A is {scm_a['Archetype']}, User B is {scm_b['Archetype']}. Interaction is {synergy['Dynamic_Type']}."
        }

if __name__ == "__main__":
    # ==========================================
    # CASE 1: SOLO USER ($10 Plan)
    # ==========================================
    print("--- SOLO REPORT GENERATION ---")
    solo_user = SoloIdentityEngine(
        p=85, s=90, v=80, t=70,  # ハイテンションな声
        mbti_type="INFP"         # でも本人は内向的と主張
    )
    solo_payload = solo_user.generate_payload()
    print(solo_payload)
    # Output Example: 
    # Archetype: The Lightning Rod
    # Gap: Over-Amplified (Masking Introversion) -> LLM will roast this.


    # ==========================================
    # CASE 2: COUPLE USERS ($15 Plan)
    # ==========================================
    print("\n--- COUPLE REPORT GENERATION ---")
    tom = {'name': 'Tom', 'job': 'lawyer', 'accent': 'NY', 'p': 30, 's': 80, 'v': 85, 't': 40}
    mary = {'name': 'Mary', 'job': 'artist', 'accent': 'LA', 'p': 80, 's': 40, 'v': 40, 't': 80}
    
    couple_engine = CoupleResonanceEngine(tom, mary)
    couple_payload = couple_engine.generate_payload()
    print(couple_payload)
    # Output Example:
    # Tom: Efficient Strategist (High Comp, Low Warmth)
    # Mary: Empathetic Soul (Low Comp, High Warmth)
    # Synergy: High-Contrast / Opposites -> LLM will write about "Power vs Poetry

File: src/lib/voiceProcessor.ts
import { SOLO_AUDIT_SYSTEM_PROMPT, COUPLE_AUDIT_SYSTEM_PROMPT } from './prompts';
import { AnalysisMetrics } from './types';

/**
 * Normalize AnalysisMetrics (raw Hz/float values) to 0-100 scale
 * for use with Identity Engines.
 * 
 * Raw values from analyzer:
 * - pitch: 50-300 Hz (typical human voice)
 * - speed: 0-1 (normalized float)
 * - vibe: 0-0.5 (variance ratio, energy fluctuation)
 * - tone: 500-4000 Hz (spectral centroid)
 */
export function normalizeMetricsForEngine(metrics: AnalysisMetrics): {
    p: number; s: number; v: number; t: number;
} {
    // Pitch: 50-300 Hz → 0-100
    // Lower pitch (50Hz) = 0, Higher pitch (300Hz) = 100
    const p = Math.min(100, Math.max(0, ((metrics.pitch - 50) / 250) * 100));

    // Speed: 0-1 → 0-100
    const s = Math.min(100, Math.max(0, metrics.speed * 100));

    // Volume: Use 'vibe' (variance) as proxy for volume/energy
    // Higher variance = more dynamic/loud voice
    // vibe typically ranges 0-0.5 → map to 0-100
    const v = Math.min(100, Math.max(0, (metrics.vibe / 0.5) * 100));

    // Tone: 500-4000 Hz → 0-100
    // Lower tone (500Hz) = 0 (husky/deep), Higher tone (4000Hz) = 100 (clear/bright)
    const t = Math.min(100, Math.max(0, ((metrics.tone - 500) / 3500) * 100));

    return {
        p: Math.round(p),
        s: Math.round(s),
        v: Math.round(v),
        t: Math.round(t)
    };
}

// ==========================================
// 1. UTILS: Demographic Mapping
// ==========================================
export type AgeGroup = 'gen_z' | '20s' | '30s' | '40s' | '50s' | '60+';

export function getAgeGroup(birthYear: number): AgeGroup {
    const currentYear = new Date().getFullYear();
    const age = currentYear - birthYear;
    if (age < 20) return 'gen_z';
    if (age < 30) return '20s';
    if (age < 40) return '30s';
    if (age < 50) return '40s';
    if (age < 60) return '50s';
    return '60+';
}

// ==========================================
// 2. CORE: Demographic Normalizer
// ==========================================
class DemographicNormalizer {
    private static AGE_FACTORS: Record<AgeGroup, { s: number; v: number; p: number }> = {
        'gen_z': { s: 1.1, v: 1.0, p: 1.0 },
        '20s': { s: 1.0, v: 1.0, p: 1.0 },
        '30s': { s: 1.0, v: 1.0, p: 1.0 },
        '40s': { s: 0.95, v: 1.0, p: 1.0 },
        '50s': { s: 0.9, v: 0.95, p: 0.95 },
        '60+': { s: 0.8, v: 0.9, p: 0.9 }
    };

    private static GENDER_PITCH_BASE: Record<string, number> = {
        'male': 30,
        'female': 65,
        'non-binary': 50,
        'other': 50
    };

    constructor(
        private p: number,
        private s: number,
        private v: number,
        private t: number,
        private gender: string = 'other',
        private ageGroup: AgeGroup = '30s'
    ) { }

    public normalize() {
        // 1. Age Adjustment (Speed & Volume)
        const factors = DemographicNormalizer.AGE_FACTORS[this.ageGroup] || DemographicNormalizer.AGE_FACTORS['30s'];
        const normS = this.s / factors.s;
        const normV = this.v / factors.v;

        // 2. Gender Adjustment (Pitch)
        const baseP = DemographicNormalizer.GENDER_PITCH_BASE[this.gender.toLowerCase()] || 50;
        const normP = (this.p - baseP) + 50;

        // Tone is kept mostly raw for now
        const normT = this.t;

        return {
            p: Math.floor(Math.max(0, Math.min(100, normP))),
            s: Math.floor(Math.max(0, Math.min(100, normS))),
            v: Math.floor(Math.max(0, Math.min(100, normV))),
            t: Math.floor(Math.max(0, Math.min(100, normT)))
        };
    }
}

// ==========================================
// 3. SHARED CORE: Acoustic Quantizer
// ==========================================
class AcousticQuantizer {
    static getPitchTag(value: number): string {
        if (value < 20) return "Sub-bass / Gravitas (重厚)";
        if (value < 40) return "Deep / Resonant (共鳴)";
        if (value < 60) return "Baritone / Grounded (安定的)";
        if (value < 80) return "Tenor / Clear (明瞭)";
        return "Soprano / Crystalline (透き通った)";
    }

    static getSpeedTag(value: number): string {
        if (value < 20) return "Largo / Contemplative (熟考)";
        if (value < 40) return "Andante / Deliberate (慎重)";
        if (value < 60) return "Moderato / Conversational (会話的)";
        if (value < 80) return "Allegro / Energetic (活発)";
        return "Presto / Urgent (性急)";
    }

    static getVolumeTag(value: number): string {
        if (value < 20) return "Whisper / Intimate (親密・秘密)";
        if (value < 40) return "Soft / Gentle (穏やか・配慮)";
        if (value < 60) return "Mezzo / Balanced (バランス)";
        if (value < 80) return "Forte / Projecting (発信・主張)";
        return "Fortissimo / Commanding (威厳・支配)";
    }

    static getToneTag(value: number): string {
        if (value < 20) return "Husky / Textured (陰影・ハスキー)";
        if (value < 40) return "Smoky / Warm (温かみ・スモーキー)";
        if (value < 60) return "Neutral / Natural (自然的)";
        if (value < 80) return "Bright / Polished (磨かれた)";
        return "Piercing / Pure (純粋・鋭い)";
    }

    static getAllTags(p: number, s: number, v: number, t: number) {
        return {
            "Pitch_Tag": this.getPitchTag(p),
            "Speed_Tag": this.getSpeedTag(s),
            "Volume_Tag": this.getVolumeTag(v),
            "Tone_Tag": this.getToneTag(t)
        };
    }
}

// ==========================================
// 4. SOLO ENGINE: Voice Identity & Gap Analysis
// ==========================================
export class SoloIdentityEngine {
    private stats: { p: number; s: number; v: number; t: number };
    private ageGroup: AgeGroup;

    constructor(
        private rawP: number,
        private rawS: number,
        private rawV: number,
        private rawT: number,
        private mbti: string,
        private gender: string = 'other',
        private birthYear?: number
    ) {
        this.mbti = mbti.toUpperCase();
        this.ageGroup = birthYear ? getAgeGroup(birthYear) : '30s';

        const normalizer = new DemographicNormalizer(rawP, rawS, rawV, rawT, gender, this.ageGroup);
        this.stats = normalizer.normalize();
    }

    private calculateAxes() {
        const projection = (this.stats.v + this.stats.s) / 2;
        const texture = (this.stats.p + this.stats.t) / 2;
        return { projection, texture };
    }

    private determineArchetype(proj: number, text: number) {
        if (proj >= 50) {
            if (text >= 50) {
                return {
                    Label: "The Lightning Rod (雷撃の扇動者)",
                    Quote: "Words like electric shocks. Impossible to ignore.",
                    Vibe: "High Voltage & Sharp"
                };
            } else {
                return {
                    Label: "The Thunder King (轟く覇者)",
                    Quote: "A voice that shakes the floorboards. Pure dominance.",
                    Vibe: "High Voltage & Heavy"
                };
            }
        } else {
            if (text >= 50) {
                return {
                    Label: "The Ice Sculptor (氷の彫刻家)",
                    Quote: "Precision over volume. Every syllable cuts deep.",
                    Vibe: "Low Voltage & Sharp"
                };
            } else {
                return {
                    Label: "The Midnight FM (深夜のDJ)",
                    Quote: "Velvet frequencies. You bypass ears and speak to the soul.",
                    Vibe: "Low Voltage & Heavy"
                };
            }
        }
    }

    private analyzeGap(proj: number, text: number) {
        const expectedProj = this.mbti.includes('E') ? 80 : 30;
        const expectedText = this.mbti.includes('T') ? 80 : 30;

        const projGap = proj - expectedProj;
        const textGap = text - expectedText;

        const tags: string[] = [];

        if (projGap > 30) tags.push("Over-Amplified (Masking Introversion)");
        else if (projGap < -30) tags.push("Under-Projecting (Holding Back)");
        else tags.push("Authentic Projection (Aligned)");

        if (textGap > 30) tags.push("Cooler than Personality (Logical Mask)");
        else if (textGap < -30) tags.push("Warmer than Personality (Social Mask)");
        else tags.push("Authentic Texture (Aligned)");

        return {
            Projection_Delta: Math.floor(projGap),
            Texture_Delta: Math.floor(textGap),
            Diagnosis_Tags: tags
        };
    }

    public generatePayload() {
        const { projection, texture } = this.calculateAxes();
        const archetype = this.determineArchetype(projection, texture);
        const gapAnalysis = this.analyzeGap(projection, texture);
        const quantizedMetrics = AcousticQuantizer.getAllTags(this.stats.p, this.stats.s, this.stats.v, this.stats.t);

        return {
            User_Profile: {
                MBTI: this.mbti,
                Gender: this.gender,
                Age_Group: this.ageGroup
            },
            Voice_Archetype: {
                Label: archetype.Label,
                Quote: archetype.Quote,
                Stats: { Projection: Math.floor(projection), Texture: Math.floor(texture) }
            },
            Gap_Analysis: gapAnalysis,
            Acoustic_Tags: quantizedMetrics,
            System_Instruction: `Analyze the user as a ${this.ageGroup} ${this.gender}. Focus on the discrepancy between MBTI and Voice Archetype.`
        };
    }
}

// ==========================================
// 5. COUPLE ENGINE: Resonance & SCM Analysis
// ==========================================
interface UserData {
    name: string;
    job: string;
    accent: string;
    p: number;
    s: number;
    v: number;
    t: number;
    gender?: string;
    birthYear?: number;
}

export class CoupleResonanceEngine {
    private static JOB_DB: Record<string, [number, number]> = {
        "lawyer": [90, 30], "executive": [90, 20], "engineer": [85, 40],
        "doctor": [85, 70], "founder": [85, 65], "consultant": [80, 40],
        "artist": [50, 90], "teacher": [60, 85], "designer": [60, 80],
        "nurse": [50, 95], "writer": [40, 80], "musician": [45, 90],
        "student": [50, 60], "sales": [70, 80], "other": [50, 50]
    };

    private statsA: { p: number; s: number; v: number; t: number };
    private statsB: { p: number; s: number; v: number; t: number };

    constructor(private ua: UserData, private ub: UserData) {
        const normA = new DemographicNormalizer(ua.p, ua.s, ua.v, ua.t, ua.gender, ua.birthYear ? getAgeGroup(ua.birthYear) : '30s');
        this.statsA = normA.normalize();

        const normB = new DemographicNormalizer(ub.p, ub.s, ub.v, ub.t, ub.gender, ub.birthYear ? getAgeGroup(ub.birthYear) : '30s');
        this.statsB = normB.normalize();
    }

    private calculateSCM(job: string, stats: { p: number; s: number; v: number; t: number }) {
        const [baseC, baseW] = CoupleResonanceEngine.JOB_DB[job.toLowerCase()] || [50, 50];

        const voiceCImpact = ((stats.s + stats.v) / 2 - 50) * 0.5;
        const pFactor = stats.p < 85 ? stats.p : (170 - stats.p);
        const voiceWImpact = ((stats.t + pFactor) / 2 - 50) * 0.5;

        const finalC = Math.max(0, Math.min(100, baseC + voiceCImpact));
        const finalW = Math.max(0, Math.min(100, baseW + voiceWImpact));

        let label = "The Free Spirit (Unconventional)";
        if (finalC >= 50 && finalW >= 50) label = "The Charismatic Ideal (Admiration)";
        else if (finalC >= 50 && finalW < 50) label = "The Efficient Strategist (Respect)";
        else if (finalC < 50 && finalW >= 50) label = "The Empathetic Soul (Sympathy)";

        return { Competence: Math.floor(finalC), Warmth: Math.floor(finalW), Archetype: label };
    }

    private calculateSynergy() {
        const deltaP = Math.abs(this.statsA.p - this.statsB.p);
        const deltaS = Math.abs(this.statsA.s - this.statsB.s);
        const deltaV = Math.abs(this.statsA.v - this.statsB.v);
        const deltaT = Math.abs(this.statsA.t - this.statsB.t);

        const meanDelta = (deltaP + deltaS + deltaV + deltaT) / 4;
        const syncScore = 100 - meanDelta;

        let dtype = "Complementary / Balanced";
        if (syncScore > 80) dtype = "High-Sync / Mirroring";
        else if (syncScore < 40) dtype = "High-Contrast / Opposites";

        return {
            Sync_Score: Math.floor(syncScore),
            Mean_Delta: Math.floor(meanDelta),
            Dynamic_Type: dtype,
            Volume_Dominance_A: this.statsA.v > this.statsB.v
        };
    }

    public generatePayload() {
        const scmA = this.calculateSCM(this.ua.job, this.statsA);
        const tagsA = AcousticQuantizer.getAllTags(this.statsA.p, this.statsA.s, this.statsA.v, this.statsA.t);

        const scmB = this.calculateSCM(this.ub.job, this.statsB);
        const tagsB = AcousticQuantizer.getAllTags(this.statsB.p, this.statsB.s, this.statsB.v, this.statsB.t);

        const synergy = this.calculateSynergy();

        return {
            Report_Type: "Couple_Resonance_v1",
            Relationship_Core: synergy,
            User_A_Insight: {
                Name: this.ua.name,
                Profile: `${this.ua.job} (${this.ua.gender} ${this.ua.birthYear ? new Date().getFullYear() - this.ua.birthYear : ''})`,
                SCM_Profile: scmA,
                Acoustic_Tags: tagsA
            },
            User_B_Insight: {
                Name: this.ub.name,
                Profile: `${this.ub.job} (${this.ub.gender} ${this.ub.birthYear ? new Date().getFullYear() - this.ub.birthYear : ''})`,
                SCM_Profile: scmB,
                Acoustic_Tags: tagsB
            },
            Narrative_Hint: `Interaction is ${synergy.Dynamic_Type}.`
        };
    }
}
